src/MC/fix_rmc_partial.cpp

```diff
@@ -0,0 +1,1381 @@
+  type_dist = std::uniform_int_distribution<>(0, num_charge_states - 1);
+  acc_dist = std::uniform_real_distribution<>(0, 1);
+
+  // Check if we are doing dihedral/angle/bond modifications
+  if (rmcdata.contains("dihedral_modification")) {
+    if (std::string(rmcdata["dihedral_modification"]) == "y") {
+      do_dihedral = 1;
+      if (!rmcdata.contains("dihedral_types") || !rmcdata.contains("dihedral_list")) {
+        error->all(
+            FLERR, 3,
+            "The dihedral_types and/or dihedral_list field are missing in the JSON input file");
+      }
+    } else {
+      do_dihedral = 0;
+    }
+  } else {
+    do_dihedral = 0;
+  }
+
+  if (rmcdata.contains("angle_modification")) {
+    if (std::string(rmcdata["angle_modification"]) == "y") {
+      do_angle = 1;
+      if (!rmcdata.contains("angle_types") || !rmcdata.contains("angle_list")) {
+        error->all(FLERR, 3,
+                   "The angle_types and/or angle_list field are missing in the JSON input file");
+      }
+    } else {
+      do_angle = 0;
+    }
+  } else {
+    do_angle = 0;
+  }
+
+  if (rmcdata.contains("bond_modification")) {
+    if (std::string(rmcdata["bond_modification"]) == "y") {
+      do_bond = 1;
+      if (!rmcdata.contains("bond_types") || !rmcdata.contains("bond_list")) {
+        error->all(FLERR, 3,
+                   "The bond_types and/or bond_list field are missing in the JSON input file");
+      }
+    } else {
+      do_bond = 0;
+    }
+  } else {
+    do_bond = 0;
+  }
+
+  // Read in the dihedrals that need to be altered
+  if (do_dihedral == 1) {
+    num_dihedrals = rmcdata["dihedral_list"].size();
+    dihedral_types = new int[num_charge_states];
+    if (rmcdata["dihedral_types"].size() != num_charge_states) {
+      error->all(FLERR, 3, "Number of dihedral types doesn't match the number of charge states\n");
+    }
+    for (int i = 0; i < num_charge_states; i++) {
+      dihedral_types[i] = rmcdata["dihedral_types"][i];
+    }
+    dihedral_list = new int *[num_dihedrals];
+    for (int i = 0; i < num_dihedrals; i++) {
+      dihedral_list[i] = new int[5];
+      for (int j = 0; j < 4; j++) dihedral_list[i][j] = rmcdata["dihedral_list"][i][j];
+      dihedral_list[i][4] = determine_molecule(dihedral_list[i][0]);
+    }
+  }
+
+  // Read in the angles that need to be altered
+  if (do_angle == 1) {
+    num_angles = rmcdata["angle_list"].size();
+    angle_types = new int[num_charge_states];
+    if (rmcdata["angle_types"].size() != num_charge_states) {
+      error->all(FLERR, 3, "Number of angle types doesn't match the number of charge states\n");
+    }
+    for (int i = 0; i < num_charge_states; i++) angle_types[i] = rmcdata["angle_types"][i];
+    angle_list = new int *[num_angles];
+    for (int i = 0; i < num_angles; i++) {
+      angle_list[i] = new int[4];
+      for (int j = 0; j < 3; j++) angle_list[i][j] = rmcdata["angle_list"][i][j];
+      angle_list[i][3] = determine_molecule(angle_list[i][0]);
```

Comment:
The original code had an `angle_list[i][4]` here, which is clearly an out-of-bounds access. This may have negatively impacted simulations with that code as that information would not have been stored correctly and thus some careful checking is strongly recommended.